### DISCUSSION TITLE
v0.74.0.30

### DISCUSSION THREAD

**Discussion by tillt** [tillt/KompleteSynthesia#34] [Reversing]
- adds a preliminary fuzzing dialog for bruteforcing the lightguide and other control commands for MK3 controllers

<hr /><em>This discussion was created from the release <a href='https://github.com/tillt/KompleteSynthesia/releases/tag/v0.74.0.30'>v0.74.0.30</a>.</em>

**Comment by tillt** [comment]
For testing, please start the application and then go into the menu to select `Fuzzing MK3`. That should bring up a new window.

1. leave the delay slider as is to save your valuable time
2. press `Start`
3. observe the `Current Message` text field -- it will show the entire command string sent to the controller. The first displayed value is the actual command used. That value will quickly increase and run all the way until it hits `FF`.
4. did the lightguide show any changes until `FF` was hit?
5. if YES -> press `Stop`, raise the delay with that slider to something higher - like in the middle of the range. Press `Start` again and try to see which exact value was changing anything lighting (lightguide and or button lighting). If you missed the exact point, dont worry - let it run through `FF` and overflow as it will use a new color in the next round, making the change obvious.
6. if NO -> darn, we didn't find anything useful this way

Any command value that had any visible effect on the controller is interesting to me.

If run with an MK2 controller, we will see that `80` changes button lighting, `81` changes lightguide, `A0` does something as well.

**Comment by pr0jectile** [comment]
Hi @tillt decided to run the fuzzing in **v0.74.0.41** for over an hour to see how it went as I had some free time today.

I noticed it was numbers like `81 XX XX XX...`, `82 XX XX XX...`, `83 XX XX XX...` and `F4 XX XX XX...` that changed lights but not at every run after FF.

I have recorded a video which I am putting together with a screen recording and a phone recording of the S88 MK3.

But for now just to give you some info, when the keyboard is plugged into the computer and Komplete Synthesia is running (I had Synthesia opened in the background but not Native Instruments or Kontact, not sure if this helps).

Ran the fuzzing and the keyboard started off with all the lights off, just the LCD screen was running throughout.

First sign of lights coming on with colour red was at:

```
81 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06
```

This turned on or changed the lights to red on the navigation buttons above the LCD screen and the dial to the right of the LCD screen.

Next one was:

```
82 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06
```

This turned on or changed the lights to red on the touch strip on the left, the 3 buttons above the strip (fixed vel and octave buttons) and wheel lights.

```
83 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06
```

This turned on or changed the lights to red on the piano key lights

```
F4 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06 06
```

Not sure if this was a glitch but this dimmed all the lights (not sure by how much but to my eyes they looked like 50-60% dimmer). Sometimes this `F4` or `F5` (not sure which) slightly increased the brightness but not back to the max brightness.

Next batch was yellows:

```
81 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12 12
```

Here are the colours I noticed that came on over the batches from the start:

- Red
- Yellow
- Green
- Blue
- Purple
- Pink
- White (for a while during the iterations, there was a whole period in the middle of the fuzzing where the keys were just white)
- Orange
- Yellow
- Torqouise
- Blue
- Pink/Purple
- White (til the end and the fuzzing started again)

I'll try and edit this video version so you can see it better if that helps (just speeding up non-action areas at the moment on it).

Hope the above helps for now!

**Reply by tillt** [reply]
Hey @pr0jectile - great results as they confirm what we saw earlier.  https://github.com/tillt/KompleteSynthesia/discussions/29#discussioncomment-8262436

We are able to trigger light guide on various zones of the controller. The biggest piece missing is how to do that while still being able to trigger MIDI sends by the controller.

**Reply by pr0jectile** [reply]
Thanks for the response – I did some debugging of the original python script and found that the key light strip is using 0x82, same as MK1! I just don't know how to make the keys work now lol. Hope this helps!

**Reply by tillt** [reply]
not really :D .... no news then so far.

**Reply by pr0jectile** [reply]
Hmm, do you reckon the HID access into the controller is taking over the device on macOS? As soon as you launch the script the entire keyboard stops working, only the LED messages. Same thing was happening with the original python script after modifying the sysex values directly to reflect MK3.
